### ISSUE TITLE
Oj.dump(mode: :rails) and Oj::Rails.encode is not the same

### ISSUE THREAD

**Issue by stereobooster** [ohler55/oj#371]
```
bundle exec ruby compatibility_test.rb
Comparing Rails to_json with other JSON implementations
+---------------------------------+--------------+---------+------------------+------------+-----------------+
| class                           | JSON to_json | Oj.dump | Oj::Rails.encode | Oj to_json | msgpack "rails" |
+---------------------------------+--------------+---------+------------------+------------+-----------------+
| Regexp                          | 👌           | 👌      | 👌               | 👌         | 👌              |
| FalseClass                      | 👌           | 👌      | 👌               | 👌         | 👌              |
| NilClass                        | 👌           | 👌      | 👌               | 👌         | 👌              |
| Object                          | ❌           | 👌      | 👌               | 👌         | 👌              |
| TrueClass                       | 👌           | 👌      | 👌               | 👌         | 👌              |
| String                          | 👌           | 👌      | 👌               | 👌         | 👌              |
| StringChinese                   | 👌           | 👌      | 👌               | 👌         | 👌              |
| StringSpecial                   | ❌           | ❌      | 👌               | 👌         | 👌              |
| StringSpecial2                  | 👌           | 👌      | 👌               | 👌         | 👌              |
| StringSpecial3                  | 👌           | 👌      | 👌               | 👌         | 👌              |
| Numeric                         | 👌           | 👌      | 👌               | 👌         | 👌              |
| Symbol                          | 👌           | 👌      | 👌               | 👌         | 👌              |
| Time                            | ❌           | 👌      | 👌               | 👌         | 👌              |
| Array                           | 👌           | 👌      | 👌               | 👌         | 👌              |
| Hash                            | 👌           | 👌      | 👌               | 👌         | 👌              |
| HashNotEmpty                    | 👌           | 👌      | 👌               | 👌         | 👌              |
| Date                            | 👌           | 👌      | 👌               | 👌         | 👌              |
| DateTime                        | ❌           | 👌      | 👌               | 👌         | 👌              |
| Enumerable                      | ❌           | 👌      | 👌               | 👌         | 👌              |
| BigDecimal                      | ❌           | 👌      | 👌               | 👌         | 👌              |
| BigDecimalInfinity              | ❌           | 👌      | 👌               | 👌         | 👌              |
| Struct                          | ❌           | 👌      | 👌               | 👌         | 👌              |
| Float                           | 👌           | 👌      | 👌               | 👌         | 👌              |
| FloatInfinity                   | 💀           | 👌      | 👌               | 👌         | ❌              |
| Range                           | 👌           | 👌      | 👌               | 👌         | 👌              |
| Complex                         | 👌           | ❌      | ❌               | 👌         | 👌              |
| Exception                       | ❌           | 👌      | 👌               | 👌         | 👌              |
| OpenStruct                      | ❌           | 👌      | 👌               | 👌         | 👌              |
| Rational                        | 👌           | ❌      | ❌               | 👌         | 👌              |
| Process::Status                 | ❌           | 👌      | 👌               | 👌         | 👌              |
| ActiveSupport::TimeWithZone     | ❌           | 👌      | 👌               | 👌         | 👌              |
| ActiveModel::Errors             | ❌           | 👌      | 👌               | 👌         | 👌              |
| ActiveSupport::Duration         | ❌           | 👌      | 👌               | 👌         | 👌              |
| ActiveSupport::Multibyte::Chars | ❌           | 👌      | 👌               | 👌         | 👌              |
| ActiveRecord::Relation          | ❌           | 👌      | 👌               | 👌         | 👌              |
| ActiveRecord                    | ❌           | 👌      | 👌               | 👌         | 👌              |
+---------------------------------+--------------+---------+------------------+------------+-----------------+
```

**Comment by ohler55** [comment]
You are halfway correct. Oj was tested agains Rails and against the json gem but not with the two together. The two lines at the top of helper.rb
```
require 'json/add/complex'
require 'json/add/rational'
```
changes the behavior of the json gem. To be consistent this should be used with Oj as described in the mimic JSON doc pages.
```
Oj.add_to_json(Complex, Rational)
```
But that does not change the Rails mimicking and it should. I'll get that changed and add some tests for that. I suspect there will be other cases where the the combination of the json gem and Rails gives mismatched results when compared to Oj mimicking.

**Comment by stereobooster** [comment]
I'm actually more worried about `StringSpecial` case  Oj.dump vs Oj::Rails.encode.

Concerning `Oj.add_to_json` interface it is counter intuitive: end user doesn't know what magic words to put into this for it to work. I more expected function: `Oj.optimise_rails()` instead of `Oj::Rails.optimize(Array, BigDecimal, Hash, Range, Regexp, Time)`. If I put `Oj::Rails.optimize(DateTime)` it will actually break encoding.

**Comment by ohler55** [comment]
The way the json gem activates the additions is with require of a file. I can't really get around that without some have that convinces ruby files are not in the place it expects them to be in a completely under the covers hack that would probably break other code. The Oj.add_to_json seemed like a reasonable alternative. Do you have a better suggestion?

I could easily add an alias so Oj.optimize_rails() call Oj::Rails.optimize() but I don't see a real advantage there. If you are referring to naming the classes then note that Oj::Rails.optimize() can be called with no arguments. That will optimize all default and previously optimized. Right now only Hash, Array, BigDecimal, Range, RegExp, and Time are optimized. All classes can not be optimized by default as there is special code from each class. Eventually I should be able to get most but for this first cut only the 4 mentioned are supported. So, a bit about why you want to be able to turn optimization on an off for a class. ActiveSupport allows the as_json method to be over-ridden but with optimization on the as_json is ignored in favor of the optimized version so if you want the as_json method called optimization needs to be turned off for that class.

The above is why DateTime fails with optimization. There is a default that might work but it does not for DateTime.

**Comment by ohler55** [comment]
The undef branch has a fix that make all Oj tests to pass as long as the ```Oj.add_to_json(Complex, Rational)``` is added.

**Comment by ohler55** [comment]
I released 3.0.3 with the changes and some other fixes.

**Comment by stereobooster** [comment]
@ohler55 yes it fixes `StringSpecial` case Oj.dump vs Oj::Rails.encode. Thanks
